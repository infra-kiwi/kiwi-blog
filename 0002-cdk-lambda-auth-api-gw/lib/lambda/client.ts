/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { Handler } from 'aws-lambda';
import { createSignedFetcher } from 'aws-sigv4-fetch';
import { inspect } from 'node:util';

const { SERVER_GATEWAY_REGION, SERVER_GATEWAY_URL } = process.env;

if (SERVER_GATEWAY_REGION == null) {
  throw new Error('SERVER_GATEWAY_REGION is required');
}
if (SERVER_GATEWAY_URL == null) {
  throw new Error('SERVER_GATEWAY_URL is required');
}

// Create the signer resource, which will take care of authenticating
// our requests to the Server API Gateway
const signedFetcher = createSignedFetcher({
  service: 'execute-api',
  region: SERVER_GATEWAY_REGION
});

export const handler: Handler = async (event) => {
  console.log(`Got event: ${event.body}`);

  // Tell the Server Lambda our name!
  const requestBody = {
    name: 'Mario'
  };

  // Strip any possible trailing slash
  const gatewayUrl = SERVER_GATEWAY_URL.replace(/\/$/, '');

  // Perform the request to the Server API Gateway
  const response = await signedFetcher(`${gatewayUrl}/my-lambda`, {
    body: JSON.stringify(requestBody),
    headers: { 'Content-Type': 'application/json' },
    method: 'POST'
  });

  const responseBody = await response.text();
  console.log(`Response: ${inspect(responseBody, { depth: null })}`);

  if (response.status != 200) {
    throw new Error(`Bad response code: ${response.status}`);
  }

  // This will let us know what we sent and what came back
  return {
    request: requestBody,
    response: responseBody
  };
};
