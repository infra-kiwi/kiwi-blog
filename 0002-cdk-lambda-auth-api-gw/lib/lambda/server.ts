import { APIGatewayProxyHandler } from 'aws-lambda';
import { inspect } from 'node:util';

export const handler: APIGatewayProxyHandler = async (event) => {
  console.log(`Got event: ${inspect(event, { depth: null })}`);

  const body = JSON.parse(event.body ?? '{}');
  const { name } = body;
  if (name) {
    return {
      statusCode: 200,
      body: `Hello ${name}!`
    };
  }

  return {
    statusCode: 401,
    body: "I don't know your name!"
  };
};
