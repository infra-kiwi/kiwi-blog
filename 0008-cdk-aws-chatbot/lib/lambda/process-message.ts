import { PublishCommand, SNSClient } from '@aws-sdk/client-sns';
import * as util from 'node:util';
import {
  ChatBotPayload,
  SlackMessagePayload,
  SNSEvent
} from './process-message.types';

const { SNS_TOPIC_ARN } = process.env;
if (!SNS_TOPIC_ARN) {
  throw new Error('Missing SNS_TOPIC_ARN env var');
}

const snsClient = new SNSClient({
  region: SNS_TOPIC_ARN.split(':')[3]
});

export const handler = async (event: SNSEvent) => {
  console.log(`Received event: ${util.inspect(event, { depth: null })}`);

  // https://aws.amazon.com/sns/faqs/
  // Q: Will a notification contain more than one message?
  // No, all notification messages will contain a single published message.
  const record = event.Records[0];

  return await processSNSRecord(record);
};

async function processSNSRecord(record: SNSEvent['Records'][number]) {
  // If we have a subject, expect a plan subject/message string pair
  if (record.Sns.Subject) {
    return await sendMessage({
      title: record.Sns.Subject,
      body: record.Sns.Message
    });
  }

  // Map the SNS message to our custom payload
  let payload: SlackMessagePayload;

  // Try to treat the SNS message as a JSON-encoded string,
  // otherwise treat it as a plain message
  try {
    const parsed = JSON.parse(record.Sns.Message);
    if (!('body' in parsed)) {
      throw new Error('Not a valid message payload');
    }
    payload = parsed;
  } catch (err) {
    console.debug('Failed to JSON-parse SNS message', err);
    payload = {
      body: record.Sns.Message
    };
  }

  return await sendMessage(payload);
}

async function sendMessage(payload: SlackMessagePayload) {
  const { title, body } = payload;

  const message: ChatBotPayload = {
    version: '1.0',
    source: 'custom',
    content: {
      title,
      description: body
    }
  };

  console.log(
    `Publishing message to ${SNS_TOPIC_ARN}: ${util.inspect(message, { depth: null })}`
  );

  return await snsClient.send(
    new PublishCommand({
      TopicArn: SNS_TOPIC_ARN,
      Message: JSON.stringify(message)
    })
  );
}
