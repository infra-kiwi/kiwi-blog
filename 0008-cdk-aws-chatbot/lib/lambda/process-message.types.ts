// Our custom payload format
export interface SlackMessagePayload {
  // The title of the Slack message, optional
  title?: string;

  // The body of the Slack message
  body: string;
}

// The payload that SNS sends to Lambdas
// https://docs.aws.amazon.com/lambda/latest/dg/with-sns.html
export interface SNSEvent {
  Records: {
    EventSource: string;
    EventVersion: string;
    EventSubscriptionArn: string;
    Sns: {
      Type: string;
      TopicArn: string;
      Subject?: string;
      Message: string;
      Timestamp: string;
    };
  }[];
}

// The ChatBot message payload.
// You can find more info about the payload format at
// https://docs.aws.amazon.com/chatbot/latest/adminguide/custom-notifs.html
export interface ChatBotPayload {
  version: '1.0';
  source: 'custom';
  content: {
    title?: string;
    description: string;
  };
}
