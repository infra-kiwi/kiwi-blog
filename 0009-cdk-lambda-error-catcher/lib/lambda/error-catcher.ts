/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import {
  CloudWatchLogsDecodedData,
  CloudWatchLogsHandler,
  CloudWatchLogsLogEvent
} from 'aws-lambda';
import * as util from 'node:util';
import { inspect, promisify } from 'node:util';
import { gunzip } from 'node:zlib';
import { PublishCommand, SNSClient } from '@aws-sdk/client-sns';

// Expects an SNS topic that accepts a {title,body} payload, like the one configured in
// https://blog.infra.kiwi/aws-cdk-a-slack-notification-pipeline-via-aws-chatbot-0e0c76e7f4c3
const { ERRORS_SNS_TOPIC_ARN, AWS_REGION } = process.env;

const gunzipPromise = promisify(gunzip);

const snsClient = ERRORS_SNS_TOPIC_ARN
  ? new SNSClient({
      region: ERRORS_SNS_TOPIC_ARN.split(':')[3]
    })
  : undefined;

export const handler: CloudWatchLogsHandler = async function (event) {
  console.log(`Received event: ${inspect(event, { depth: null })}`);

  // Decode the CloudWatch payload, which comes gzipped
  const payloadBase64 = Buffer.from(event.awslogs.data, 'base64');
  const payloadString = (await gunzipPromise(payloadBase64)).toString('utf-8');
  const payload = JSON.parse(payloadString) as CloudWatchLogsDecodedData;

  console.log(
    `Received error event: ${util.inspect(payload, { depth: null })}`
  );

  if (snsClient) {
    // Prettify the log group name
    const logGroupName = payload.logGroup.replace(/^\/aws\/lambda\//, '');

    // Build the logs url, so we can directly access the correct
    // CloudWatch page. This url makes use of some double encoding, sorry
    // for the looks, but that's how CloudWatch urls look like!
    const logsUrlPrefix = [
      `https://${AWS_REGION}.console.aws.amazon.com/cloudwatch/home?region=${AWS_REGION}`,
      `#logsV2:log-groups/log-group/${cloudWatchUrlEncode(encodeURIComponent(payload.logGroup))}`,
      `/log-events/${cloudWatchUrlEncode(encodeURIComponent(payload.logStream))}`
    ].join('');

    // Prettify the errors
    const errorsString = payload.logEvents
      .map((evt) => prettifyError(evt, logsUrlPrefix))
      .join('\n\n---\n\n');

    const notificationPayload = {
      title: `ðŸš¨ Lambda error: ${logGroupName}`,
      body: `*${payload.owner}* - The following errors were detected:\n\n${errorsString}`
    };
    await snsClient.send(
      new PublishCommand({
        TopicArn: ERRORS_SNS_TOPIC_ARN,
        Message: JSON.stringify(notificationPayload)
      })
    );
  }
};

export function prettifyError(
  event: CloudWatchLogsLogEvent,
  logsUrlPrefix: string
): string {
  let details:
    | {
        timestamp: string;
        message: string;
      }
    | undefined;

  // Evaluate if the message content is a JSON object
  if (event.message.startsWith('{')) {
    try {
      // If we receive a JSON error, we can decode it and extract the real
      // error message
      const payload = JSON.parse(event.message);

      const timestamp = payload.timestamp;
      let message = payload.message;

      if (timestamp == null || message == null) {
        throw new Error('Unknown log payload format');
      }

      // Fallback for odd/unknown/uncovered cases
      if (typeof message !== 'string') {
        message = util.inspect(message, { depth: null });
      }

      details = {
        timestamp,
        message
      };
    } catch (err) {
      console.log('Failed to decode JSON event', err);
    }
  }

  // If there was no JSON match, try to extract the relevant data
  // using a regular expression based on the standard CloudWatch
  // TEXT logging format
  if (details?.message == null) {
    // Treat the error line as a standard TEXT line
    const match = /^([T:Z\d.-]+)\t[\w-]+\tERROR\t(.+)$/m.exec(event.message);
    if (match) {
      details = {
        timestamp: match[1],
        message: match[2]
      };
    }
  }

  // If all parsing failed, send the whole log line
  if (details?.message == null) {
    details = {
      timestamp: new Date(event.timestamp).toISOString(),
      message: event.message
    };
  }

  // Make logs start 1s before the actual log message
  // CloudWatch normally does 100ms, but it is restrictive in low
  // traffic scenarios
  const logsStartTime = event.timestamp - 1000;

  // Build the remaining part of the logs url, which will point at the right
  // time slot and highlight the current error message
  const logsUrl =
    logsUrlPrefix +
    cloudWatchUrlEncode(`?start=${logsStartTime}&refEventId=${event.id}`);

  return `
_${details.timestamp}_ (<${logsUrl}|Open CloudWatch logs>)
\`\`\`
${details.message}
\`\`\`  
`.trim();
}

// Little helper to work with CloudWatch url fragments
function cloudWatchUrlEncode(str: string) {
  return encodeURIComponent(str).replaceAll('%', '$');
}
