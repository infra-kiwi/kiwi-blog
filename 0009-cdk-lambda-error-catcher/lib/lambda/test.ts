export const handler = ({ fail }: { fail?: boolean | 'log' }) => {
  if (fail === true) {
    throw new Error('This error was thrown!');
  }

  if (fail === 'log') {
    console.error('This is an error logged via console.error');
  } else {
    console.log('All good!');
  }
};
