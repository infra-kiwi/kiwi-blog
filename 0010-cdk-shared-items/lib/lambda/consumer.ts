import { EventBridgeHandler } from 'aws-lambda';
import * as util from 'node:util';

// The whole purpose of this handler is to log the greeting
// sent by the producer Lambda
export const handler: EventBridgeHandler<'MyEvent', string, void> = async (
  event
) => {
  console.log(`Received event: ${util.inspect(event.detail, { depth: null })}`);
};
