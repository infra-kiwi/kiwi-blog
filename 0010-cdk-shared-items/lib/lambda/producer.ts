/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import {
  EventBridgeClient,
  PutEventsCommand
} from '@aws-sdk/client-eventbridge';
import * as util from 'node:util';

const { EVENT_BUS_NAME } = process.env;
if (EVENT_BUS_NAME == null) {
  throw new Error('Missing EVENT_BUS_NAME env var');
}

const eventBridgeClient = new EventBridgeClient();

// The whole purpose of this handler is to send the whole event payload
// over to the EventBus
export const handler = async (event: { greeting: string }) => {
  if (event.greeting == null) {
    throw new Error('Missing greeting!');
  }

  // Generates an event in the EventBus
  const result = await eventBridgeClient.send(
    new PutEventsCommand({
      Entries: [
        {
          Detail: JSON.stringify(event),
          DetailType: 'MyEvent',
          Source: 'ProducerLambda',
          EventBusName: EVENT_BUS_NAME
        }
      ]
    })
  );

  if ((result.FailedEntryCount ?? 0) > 0) {
    throw new Error(
      `Failed to send event: ${util.inspect(result, { depth: null })}`
    );
  }

  return { message: 'Event sent!', result };
};
