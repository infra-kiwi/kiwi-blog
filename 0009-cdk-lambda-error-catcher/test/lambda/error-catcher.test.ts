import { mockClient } from 'aws-sdk-client-mock';
import {
  PublishCommand,
  PublishCommandInput,
  PublishCommandOutput,
  SNSClient
} from '@aws-sdk/client-sns';
import { CloudWatchLogsLogEvent } from 'aws-lambda';

const snsMock = mockClient(SNSClient);

describe('error-catcher.ts Lambda', () => {
  // Prepare the environment for the Lambda function
  beforeEach(() => {
    // Fake ARN
    process.env.ERRORS_SNS_TOPIC_ARN =
      'arn:aws:sns:my-region:123456789012:chatbot-my-channel';
    process.env.AWS_REGION = 'us-east-1';

    snsMock.reset();
  });

  test('accepts a CloudWatch log message', async () => {
    const payload = require('../assets/cloudwatch_log.json');

    let snsInput: PublishCommandInput | undefined;
    snsMock.on(PublishCommand).callsFakeOnce((input: PublishCommandInput) => {
      snsInput = input;
      const result: PublishCommandOutput = {
        MessageId: 'fake',
        $metadata: {}
      };
      return result;
    });

    await require('../../lib/lambda/error-catcher').handler(payload);

    {
      expect(snsInput?.Message).not.toBeUndefined();
      expect(JSON.parse(snsInput!.Message!)).toMatchSnapshot();
    }
  });

  const prettifyCases: {
    event: CloudWatchLogsLogEvent;
    expected: string;
  }[] = [
    {
      event: require('../assets/cloudwatch_event_text.json').logEvents[0],
      expected: ''
    },
    {
      event: require('../assets/cloudwatch_event_json.json').logEvents[0],
      expected: ''
    }
  ];

  test.each(prettifyCases)('prettifies correctly the messages (%#)', (t) => {
    expect(
      require('../../lib/lambda/error-catcher').prettifyError(
        t.event,
        'https://mylogs.com'
      )
    ).toMatchSnapshot();
  });
});
