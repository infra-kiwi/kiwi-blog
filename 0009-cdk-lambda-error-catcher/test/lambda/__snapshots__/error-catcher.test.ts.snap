// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error-catcher.ts Lambda accepts a CloudWatch log message 1`] = `
{
  "body": "*123456789012* - The following errors were detected:

_2024-05-04T04:41:33.937Z_ (<https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/$252Faws$252Flambda$252Fkiwi-blog-0009-cdk-lambda-error-catchertest-lambda-json/log-events/2024$252F05$252F04$252F$255B$2524LATEST$255Db160e8ab93984ad2b4712c478779c545$3Fstart$3D1714797692937$26refEventId$3D38241266439516937806054119273176397097372920146847596545|Open CloudWatch logs>)
\`\`\`
{
  errorType: 'Error',
  errorMessage: 'This error was thrown!',
  stackTrace: [
    'Error: This error was thrown!',
    '    at Runtime.handler (/var/task/index.js:28:11)',
    '    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1173:29)'
  ]
}
\`\`\`",
  "title": "ðŸš¨ Lambda error: kiwi-blog-0009-cdk-lambda-error-catchertest-lambda-json",
}
`;

exports[`error-catcher.ts Lambda prettifies correctly the messages (0) 1`] = `
"_2024-05-04T04:47:56.437Z_ (<https://mylogs.com$3Fstart$3D1714798075437$26refEventId$3D38241274969551976244017471362271343276524099733197750274|Open CloudWatch logs>)
\`\`\`
Invoke Error 	{"errorType":"Error","errorMessage":"This error was thrown!","stack":["Error: This error was thrown!","    at Runtime.handler (/var/task/index.js:28:11)","    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1173:29)"]}
\`\`\`"
`;

exports[`error-catcher.ts Lambda prettifies correctly the messages (1) 1`] = `
"_2024-05-04T04:53:42.156Z_ (<https://mylogs.com$3Fstart$3D1714798421156$26refEventId$3D38241282679343305534825973659543398526898726925601406977|Open CloudWatch logs>)
\`\`\`
{
  errorType: 'Error',
  errorMessage: 'This error was thrown!',
  stackTrace: [
    'Error: This error was thrown!',
    '    at Runtime.handler (/var/task/index.js:28:11)',
    '    at Runtime.handleOnceNonStreaming (file:///var/runtime/index.mjs:1173:29)'
  ]
}
\`\`\`"
`;
